File: Data Structures/BST Insert and Remove/701/701.py
# Definition for a binary tree node.
class TreeNode:
    # Initialize a tree node with value, left child, and right child.
    def __init__(self, val=0, left=None, right=None):
        self.val = val  # Node's value
        self.left = left  # Left child
        self.right = right  # Right child

class Solution:
    # Function to insert a value into a binary search tree (BST) and return the root of the modified tree.
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        # If the current node is None, create a new node with the value and return it.
        # This is the base case for recursion.
        if not root:
            return TreeNode(val)
        
        # If the value to insert is greater than the current node's value,
        # insert the value in the right subtree.
        if val > root.val:
            root.right = self.insertIntoBST(root.right, val)
        else:
            # If the value is less than or equal to the current node's value,
            # insert the value in the left subtree.
            root.left = self.insertIntoBST(root.left, val)
        
        # Return the unchanged current node.
        return root
    
# Time complexity: O(H), where H is the height of the tree.
# This is because in the worst case, we have to travel from the root to the leaf.
# Space complexity: O(H) due to the recursion stack.
# In the worst case (a skewed tree), the recursion stack could be as large as the height of the tree.

File: Data Structures/Trees/BST Insert and Remove/701/701.go
package main

import "fmt"

// TreeNode represents a node in the binary search tree.
type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

// insertIntoBST inserts a new value into the BST and returns the root of the modified tree.
func insertIntoBST(root *TreeNode, val int) *TreeNode {
    // If the root is nil, the new value becomes the root.
    if root == nil {
        return &TreeNode{Val: val}
    }
    // If the value is less than the current node's value, insert it into the left subtree.
    if val < root.Val {
        root.Left = insertIntoBST(root.Left, val)
    } else {
        // Otherwise, insert it into the right subtree.
        root.Right = insertIntoBST(root.Right, val)
    }
    // Return the unchanged root node.
    return root
}

// Time complexity: O(h), where h is the height of the tree.
// Space complexity: O(h)